pkg/reconciler/workload/synctargetexports/synctargetexports_reconcile.go

```diff
@@ -0,0 +1,178 @@
+	var syncedResources []workloadv1alpha1.ResourceToSync
+	for _, exportKey := range exportKeys {
+		exportCluster, name := clusters.SplitClusterAwareKey(exportKey)
+		export, err := e.getAPIExport(exportCluster, name)
+		if apierrors.IsNotFound(err) {
+			continue
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+		for _, schema := range export.Spec.LatestResourceSchemas {
+			syncedResource, err := e.convertSchemaToSyncedResource(exportCluster, schema, export.Status.IdentityHash)
+			if err != nil {
+				klog.Warningf("cannot get schema: %v", err)
+				continue
+			}
+			syncedResources = append(syncedResources, syncedResource)
+		}
+	}
+
+	// sort synced resource by group
+	sort.SliceStable(syncedResources, func(i, j int) bool {
+		if syncedResources[i].Group > syncedResources[j].Group {
+			return true
+		}
+
+		if syncedResources[i].Resource > syncedResources[j].Resource {
+			return true
+		}
+		return false
+	})
+
+	// merge synced resource using desired as base and update it state based on existing synced.
+	for _, existingSynced := range syncTarget.Status.SyncedResources {
+		for i := range syncedResources {
+			if syncedResources[i].GroupResource == existingSynced.GroupResource && syncedResources[i].IdentityHash == existingSynced.IdentityHash {
+				syncedResources[i].State = existingSynced.State
+				break
+			}
+		}
+	}
+
+	syncTarget.Status.SyncedResources = syncedResources
+	return syncTarget, reconcileStatusContinue, errors.NewAggregate(errs)
+}
+
+func (e *exportReconciler) convertSchemaToSyncedResource(cluterName logicalcluster.Name, schemaName, identityHash string) (workloadv1alpha1.ResourceToSync, error) {
+	schema, err := e.getResourceSchema(cluterName, schemaName)
+	if err != nil {
+		return workloadv1alpha1.ResourceToSync{}, err
+	}
+
+	syncedResource := workloadv1alpha1.ResourceToSync{
+		GroupResource: apisv1alpha1.GroupResource{
+			Group:    schema.Spec.Group,
+			Resource: schema.Spec.Names.Plural,
+		},
+		Versions:     []string{},
+		IdentityHash: identityHash,
+	}
+
+	for _, version := range schema.Spec.Versions {
+		if version.Served {
+			syncedResource.Versions = append(syncedResource.Versions, version.Name)
+		}
+	}
+	sort.Strings(syncedResource.Versions)
+
+	return syncedResource, nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, syncTarget *workloadv1alpha1.SyncTarget) (*workloadv1alpha1.SyncTarget, error) {
+	reconcilers := []reconciler{
+		&exportReconciler{
+			getAPIExport:      c.getAPIExport,
+			getResourceSchema: c.getResourceSchema,
+		},
+		&apiCompatibleReconciler{
+			getAPIExport:           c.getAPIExport,
+			getResourceSchema:      c.getResourceSchema,
+			listAPIResourceImports: c.listAPIResourceImports,
+		},
+	}
+
+	var errs []error
+
+	for _, r := range reconcilers {
+		var err error
+		var status reconcileStatus
+		syncTarget, status, err = r.reconcile(ctx, syncTarget)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if status == reconcileStatusStop {
```

Comment:
Am I missing something or the only place where `reconcileStatusStop` is returned is in the `apiCompatibleReconciler` `reconcile` method ?
If yes what does it bring to test this value here and break since this reconciler is already the second one in the loop ?

More generally Is a loop of reconcilers really useful here ?
Wouldn't it be easier to read to just call both reconcilers sequentially ?

Reply:
yes, that is right. We can just call them sequentially. Updated.